Fix n/N motions after */#

zeux opened the pull request:
We need to remember the last search command for * and #, and use the command
or the reverse command on n/N.

This requires making _vi_star and _vi_octothorp act in both "grab the current
word and search" and "search for this word" modes.

zeux commented:
Currently n/N after */# use regular search commands so n after * stops using full-word mode. This change fixes it in the least invasive way.

guillermooo commented:
Looks good, thanks! We'd need tests for it. I don't think we can use the `.cmd-test` format here.

guillermooo commented on xmotions.py:
```diff
@@ -1772,16 +1778,17 @@ def on_cancel(self):
+            'vi_octothorp': ['_vi_octothorp', '_vi_star'],
+        }
+
+        command = commands[search_command][reverse]
```
I would say 

```python
 command = commands[search_command][int(reverse)]
```

guillermooo commented on xmotions.py:
```diff
@@ -1772,16 +1778,17 @@ def on_cancel(self):
-                })
-            return
+    def run(self, mode=None, count=1, search_string='', search_command='vi_slash', reverse=False):
+        commands = {
```
I think we can move this to a class attribute.

guillermooo commented:
Left a couple of minor comments. I do think we should add tests before merging this, though.

Thanks again!

zeux commented:
Are there any tests for search motions?

My understanding is that this requires a test in tests/commands/ that's similar to other tests in that it executes internal commands directly (e.g. _vi_slash), and verifies the resulting selection.

guillermooo commented:
Yes, that sounds about right. You can pick one of the newest files as an example. If you need further help, just ask!

guillermooo commented:
I'm in the process of moving all commands/motions tests to the `.cmd-test` format, but that format cannot (perhaps yet) handle a test for the search commands you've modified.

zeux commented:
I've added basic tests for */#; they seem to work.

I'm trying to figure out how to simulate n/N. Currently it's not handled with a motion command so I can't just use run_command...

zeux commented:
Huh. And by "work well" I mean "works on my Mac, works on Travis CI but fails on Appveyor" :-/

zeux commented:
I guess I can call _vi_repeat_buffer_search and remove this:

            'search_string': state.last_buffer_search,
            'search_command': state.last_buffer_search_command,

from cmd_defs.py and just use state. in the command so that the interaction between star/repeat can be tested by just executing commands in succession.

zeux commented:
Added the tests.
? actually had a bug (it did not wrap around the end if it was used from 0,0 in the buffer), I fixed that as well.

AppVeyor build still fails the tests and I'm still not sure why...

guillermooo commented:
I'll take a look at the Windows bug. Thanks!

guillermooo commented:
All tests pass for me, so probably a newline issue? I've resubmitted from my fork, where Appveyor was passing too. Let's see how it goes.

guillermooo commented:
Nope. It seems that GH has a bug and will only show the firts CI results in the branch view:

https://github.com/guillermooo-forks/Vintageous/branches

guillermooo commented:
Still, the tests pass ok locally for me...

guillermooo commented:
My hunch is that the mode in Appveyor isn't the expected one:

https://github.com/zeux/Vintageous/blob/search-star-n-fix/xmotions.py#L1240

Perhaps you have to set it explicitly in the test.

zeux commented:
For me the tests passed locally on latest Windows ST 3dev but failed on the version that AppVeyor was using (although the motions themselves worked fine in editor)

guillermooo commented:
I think you can simply do this in the tests:

```
self.state.mode = modes.NORMAL
```

guillermooo commented:
Hm. A version issue seems unlikely to me, but who knows.

zeux commented:
Yeah, it looks like the issue is with the mode.

Should the mode just always be set to normal in the test setup for all tests?

zeux commented:
Hmm, both _vi_star and _vi_question_mark compare mode *and* state.mode. Let me try to fix that instead.

zeux commented:
Yay, finally. Let me know if you want me to change anything else.

guillermooo commented:
> Should the mode just always be set to normal in the test setup for all tests?

In general, the mode should be set to the one expected by the test. In most tests, you're not testing for the mode management within Vintageous, but for how Vintageous behaves in certain mode. So it's fine to force the relevant mode, as it's a pre-requisite.

guillermooo commented:
Excellent work!

Thanks for you patience and cooperation!

zeux commented:
Note that in the latest commit I removed the mode setup - I believe it's only required because of an issue with mode/state.mode confusion in _vi_star/_vi_octothorp that I corrected in 5e69e7c (and simultaneously removed the mode setup). I can squash these commits to remove the confusing change.

guillermooo commented:
Ah, cool. LGTM.
